### PR TITLE
Cross DC gossip fixes

### PR DESCRIPTION
**Pull request by chbatey**
Adds a staggering of our probabilities and fixes the selection of the oldest nodes in a DC

Addresses https://github.com/akka/akka/issues/23803

### CODE REVIEW

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala [hunk 1]
```diff
@@ -946,7 +946,7 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
     if (isGossipSpeedupNeeded) gossip()
 
   def isGossipSpeedupNeeded: Boolean =
-    (latestGossip.overview.seen.size < latestGossip.members.size / 2)
+    latestGossip.overview.seen.size < latestGossip.members.size / 2
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: this should only count members in `selfDc`

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala [hunk 2]
```diff
@@ -102,16 +102,17 @@ import scala.util.Random
...
         case Some(set) ⇒
-          if (set.size < crossDcConnections) acc + (member.dataCenter → (set + member))
-          else acc
-        case None ⇒ acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
+          acc + (member.dataCenter → (set + member))
+        case None ⇒
+          acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
       }
-    }
+    }.mapValues(_.take(crossDcConnections))
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: `mapValues` creates a view, which we should perhaps stay away from?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/InitialMembersOfNewDcSpec.scala [hunk 3]
```diff
@@ -0,0 +1,107 @@
+package akka.cluster
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/InitialMembersOfNewDcSpec.scala]: copyright header

File: akka-cluster/src/multi-jvm/scala/akka/cluster/InitialMembersOfNewDcSpec.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+      // Check how long it takes for all other nodes to see every node as up
+      runOn(one, two, three, four) {
+        within(20.seconds) {
+          awaitAssert({
+            cluster.state.members.filter(_.status == MemberStatus.Up) should have size 4
+          })
+        }
+        val totalTime = System.nanoTime() - startTime
+        log.info("Can see new node (and all others as up): {}", Duration(totalTime, TimeUnit.NANOSECONDS).toMillis)
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/InitialMembersOfNewDcSpec.scala]: `totalTime.nanos.toMillis`

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala (hunk 2 above)
Comment (chbatey) [reply akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: Yeah i'll change it to the version that assumes that if the cluster is large the number of cross DC conns is small

Comment (chbatey) [reply akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: Also forgot to checkin the test for this, doh

**Comment by chbatey** [comment]
Updated based on your comments @patriknw and checked in the Test for the oldest members per DC

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala [hunk 5]
```diff
@@ -946,7 +946,7 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
     if (isGossipSpeedupNeeded) gossip()
 
   def isGossipSpeedupNeeded: Boolean =
-    (latestGossip.overview.seen.size < latestGossip.members.size / 2)
+    latestGossip.overview.seen.size < latestGossip.members.count(_.dataCenter == cluster.selfDataCenter) / 2
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: should also filter the `seen`

```
latestGossip.overview.seen.count(membershipState.isInSameDc)
```

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala [hunk 6]
```diff
@@ -102,16 +101,24 @@ import scala.util.Random
...
       acc.get(member.dataCenter) match {
         case Some(set) ⇒
-          if (set.size < crossDcConnections) acc + (member.dataCenter → (set + member))
-          else acc
-        case None ⇒ acc + (member.dataCenter → (SortedSet.empty(Member.ageOrdering) + member))
+          if (set.size < crossDcConnections) {
+            acc + (member.dataCenter → (set + member))
+          } else {
+            if (set.exists(member.isOlderThan)) {
+              acc + (member.dataCenter -> (set + member).take(crossDcConnections))
```
Comment (patriknw) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: good

**Review by patriknw — CHANGES_REQUESTED** [review]
one more small thing

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala (hunk 5 above)
Comment (chbatey) [reply akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: Doh

**Comment by chbatey** [comment]
Build failure was the NPE in compiler. Should be good to go now

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-cluster/src/main/resources/reference.conf [hunk 7]
```diff
@@ -235,6 +235,8 @@ akka {
 
       # The n oldest nodes in a data center will choose to gossip to another data center with
       # this probability. Must be a value between 0.0 and 1.0 where 0.0 means never, 1.0 means always.
+      # When a data center is first started (nodes < 5) a higher probability is used so other data
+      # centers find out about the new nodes more quickly
```
Comment (johanandren) [inline akka-cluster/src/main/resources/reference.conf]: D'oh. I think Patrik said I should do this in the original PR but I guess it was never done. Good!

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala [hunk 8]
```diff
@@ -970,6 +970,9 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef) extends Actor with
           else
             gossipTo(peer)
         case None ⇒ // nothing to see here
+          if (cluster.settings.Debug.VerboseGossipLogging)
+            log.debug("Cluster Node [{}] dc [{}] will not gossiping this round", selfAddress, cluster.settings.SelfDataCenter)
```
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: "will not gossip" (you're the native speaker)?

**Review by johanandren — APPROVED** [review]
LGTM with one nitpick!

File: akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala (hunk 8 above)
Comment (chbatey) [reply akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala]: I'm from the north of england, doesn't count

**Comment by johanandren** [comment]
I'm keeping fingers crossed that this will also fix #23879

**Pull request merged**